File: csharp/NedapRetailApiClient/Nedap.Retail.Api/Epcis.V1.1/TransformationEvent.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Nedap.Retail.Api.Epcis.V1._1
{
    /// <summary>
    /// A TransformationEvent captures information about an event in which one or more physical or digital objects identified by instance-level (EPC) or class-level (EPC Class) identifiers are consumed as inputs and one or more objects identified by instance-level (EPC) or class-level (EPC Class) identifiers are produced as outputs. The TransformationEvent captures the relationship between the inputs and the outputs, such that any of the inputs may have contributed in some way to each of the outputs.
    /// A TransformationEvent marks the end of the lifecycle for the input objects, and so is similar in effect to an ObjectEvent with action DELETE for the inputs.Likewise, a TransformationEvent marks the beginning of the lifecycle for the output objects, and so is similar in effect to an ObjectEvent with action ADD for the outputs.
    /// </summary>
    public class TransformationEvent : EpcisEvent
    {
        /// <summary>
        /// Constructor that sets type to object_event
        /// </summary>
        public TransformationEvent()
        {
            base.Type = "transformation_event";
        }

        /// <summary>
        /// Optional (see below). An unordered list of one or more EPCs naming specific objects to which the event pertained.
        /// </summary>
        [JsonProperty("input_epc_list")]
        public List<string> InputEpcList { get; set; }

        /// <summary>
        /// Optional (see below). An unordered list of one or more QuantityElements identifying (at the class level) objects to which the event pertained.
        /// </summary>
        [JsonProperty("input_quantity_list")]
        public List<QuantityElement> InputQuantityList { get; set; }

        /// <summary>
        /// Optional (see below). An unordered list of one or more EPCs naming specific objects to which the event pertained.
        /// </summary>
        [JsonProperty("output_epc_list")]
        public List<string> OutputEpcList { get; set; }

        /// <summary>
        /// Optional (see below). An unordered list of one or more QuantityElements identifying (at the class level) objects to which the event pertained.
        /// </summary>
        [JsonProperty("ioutput_quantity_listlmd")]
        public List<QuantityElement> OutputQuantityList { get; set; }

        /// <summary>
        /// Optional. A unique identifier that links this event to other TransformationEvents having an identical value of transformationID.
        /// </summary>
        [JsonProperty("transformation_id")]
        public string TransformationId { get; set; }

        /// <summary>
        /// Optional. The business step of which this event was a part.
        /// </summary>
        [JsonProperty("biz_step")]
        public string BizStep { get; set; }

        /// <summary>
        /// The business condition of the objects associated with the EPCs, presumed to hold true until contradicted by a subsequent event. 
        /// </summary>
        [JsonProperty("disposition")]
        public string Disposition { get; set; }

        /// <summary>
        /// The read point at which the event took place.
        /// </summary>
        [JsonProperty("read_point")]
        public string ReadPoint { get; set; }

        /// <summary>
        /// The business location where the objects associated with the EPCs may be found, until contradicted by a subsequent event.
        /// </summary>
        [JsonProperty("biz_location")]
        public string BizLocation { get; set; }

        /// <summary>
        /// Optional. An unordered list of business transactions that define the context of this event.
        /// </summary>
        [JsonProperty("biz_transaction_list")]
        public List<BusinessTransaction> BizTransactionList { get; set; }

        /// <summary>
        /// Optional. An unordered list of Source elements that provide context about the originating endpoint of a business transfer of which this event is a part.
        /// </summary>
        [JsonProperty("source_list")]
        public List<Source> SourceList { get; set; }

        /// <summary>
        /// Optional. An unordered list of Destination elements that provide context about the terminating endpoint of a business transfer of which this event is a part.
        /// </summary>
        [JsonProperty("destination_list")]
        public List<Destination> DestinationList { get; set; }

        /// <summary>
        /// Optional. Instance/Lot Master Data that describes the objects created during this event.
        /// </summary>
        [JsonProperty("ilmd")]
        public string Ilmd { get; set; }
    }
}

File: java/messages/src/main/java/com/nedap/retail/messages/epcis/v1_1/TransformationEvent.java
package com.nedap.retail.messages.epcis.v1_1;

import java.util.List;

import org.codehaus.jackson.annotate.JsonProperty;
import org.joda.time.DateTime;

import com.nedap.retail.messages.epcis.v1_1.cbv.Disposition;
import com.nedap.retail.messages.epcis.v1_1.cbv.EventType;
import com.nedap.retail.messages.epcis.v1_1.elements.QuantityElement;

public class TransformationEvent extends EpcisEvent {

    @JsonProperty("input_epc_list")
    public List<String> inputEpcList;

    @JsonProperty("output_epc_list")
    public List<String> outputEpcList;

    @JsonProperty("input_quantity_list")
    public List<QuantityElement> inputQuantityList;

    @JsonProperty("output_quantity_list")
    public List<QuantityElement> outputQuantityList;

    @JsonProperty("transformation_id")
    public String transformationId;

    public TransformationEvent() {
        type = EventType.TransformationEvent;
    }

    /**
     * An aggregation_event captures information about an event pertaining to one or more physical or digital objects
     * identified by instance-level (EPC) or class-level (EPC Class) identifiers. Most object_events are envisioned to
     * represent actual observations of objects, but strictly speaking it can be used for any event a Capturing
     * Application wants to assert about objects, including for example capturing the fact that an expected observation
     * failed to occur.
     *
     * @param id The ID that identifies this message uniquely to an organization
     * @param eventTime The date and time at which the EPCIS Capturing Applications asserts the event occurred
     * @param recordTime The date and time at which this event was recorded by the EPCIS Repository.
     * @param eventTimeZoneOffset The time zone offset in effect at the time and place the event occurred, expressed as
     *            an offset from UTC
     * @param bizLocation The business location where the objects associated with the EPCs may be found, until
     *            contradicted by a subsequent event.
     * @param readPoint The read point at which the event took place.
     * @param disposition The business condition of the objects associated with the EPCs, presumed to hold true until
     *            contradicted by a subsequent event.
     */
    public TransformationEvent(final String id, final DateTime eventTime, final DateTime recordTime,
            final String eventTimeZoneOffset, final String bizLocation, final String readPoint,
            final Disposition disposition) {
        this.id = id;
        this.eventTime = eventTime;
        this.recordTime = recordTime;
        this.eventTimeZoneOffset = eventTimeZoneOffset;
        this.type = EventType.TransformationEvent;
        this.bizLocation = bizLocation;
        this.readPoint = readPoint;
        this.disposition = disposition != null ? disposition.disposition() : Disposition.UNKNOWN.disposition();
    }

    @Override
    public String toString() {
        final String inputEpcListSize = inputEpcList == null ? "null" : Integer.toString(inputEpcList.size());
        final String outputEpcListSize = outputEpcList == null ? "null" : Integer.toString(outputEpcList.size());
        final String inputQuantityListSize = inputQuantityList == null ? "null"
                : Integer.toString(inputQuantityList.size());
        final String outputQuantityListSize = outputQuantityList == null ? "null"
                : Integer.toString(outputQuantityList.size());

        return "TransformationEvent" + super.toString() + "[input_epc_list(" + inputEpcListSize + "),output_epc_list("
                + outputEpcListSize + "),input_quantity_list(" + inputQuantityListSize + "),output_quantity_list("
                + outputQuantityListSize + ")";
    }

    @Override
    public boolean equals(final Object obj) {
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        if (!super.equals(obj)) {
            return false;
        }
        final TransformationEvent other = (TransformationEvent) obj;
        if (!compareAsSet(this.inputEpcList, other.inputEpcList)) {
            return false;
        }
        if (!compareAsSet(this.outputEpcList, other.outputEpcList)) {
            return false;
        }
        if (!compareAsSet(this.inputQuantityList, other.inputQuantityList)) {
            return false;
        }
        if (!compareAsSet(this.outputQuantityList, other.outputQuantityList)) {
            return false;
        }
        return true;
    }

    @Override
    public int hashCode() {
        int result = super.hashCode();
        result = 31 * result + (inputEpcList != null ? inputEpcList.hashCode() : 0);
        result = 31 * result + (outputEpcList != null ? outputEpcList.hashCode() : 0);
        result = 31 * result + (inputQuantityList != null ? inputQuantityList.hashCode() : 0);
        result = 31 * result + (outputQuantityList != null ? outputQuantityList.hashCode() : 0);
        result = 31 * result + (transformationId != null ? transformationId.hashCode() : 0);
        return result;
    }
}
